### PR TITLE
Render hex-typed integer fields with hex literals

### LINKED ISSUE #410: [Bug] Visual editor: i2c address field does not support hexadecimal addresses
### What happened?

- Create/Open device
- Add i2c component e.g. bme280_i2c
- The visual editor doesn't support setting the i2c address using hexadecimal as is common with i2c addresses
- Setting the address in yaml causes the visual editor to not display the correct value in the field

Image

### Dashboard version

beta24

### How did you install it?

Wheel from a GitHub release

### Where is it running?

Local — Linux

### Relevant logs / errors

```shell

```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

The frontend half of issue #410 — i2c address fields render as a
plain decimal number input. Typing `0x76` is rejected by the
browser-level number validation; setting `address: 0x76` in YAML
shows up as `118` in the form (because js-yaml parses the hex
literal to a plain integer). Both ends of the round trip lose the
hardware-conventional hex notation.

Backend companion PR esphome/device-builder#414 adds a
`display_format: "hex"` flag on the catalog entries for every
`cv.hex_uint*_t` field (~209 entries — every i2c address plus a
handful of register-address fields). This PR consumes the flag.

Adds a `renderHexIntField` branch on `renderNumberField` that
swaps the number input for a text input plus the new
`parseHexInt` / `formatHexInt` / `normalizeHexValues` helpers in
`util/hex-int.ts`:

- **Display**: any underlying numeric value reformats to `"0x" +
  lowercase-hex` so a YAML `address: 0x76` (parsed by js-yaml as
  `118`) shows as `0x76` in the form. Matches ESPHome's
  `cv.hex_int` formatter (lowercase hex).
- **Input**: accepts `0x76` / `0X76` (hex) and `118` (decimal) —
  same as ESPHome's own `cv.hex_int` validator. Unparseable input
  gets emitted as a raw string so the inline validator flags it
  instead of being silently swallowed. The editing buffer
  (mirrors the float-with-unit pattern) keeps intermediate states
  like `"0x"` / `"0x7"` intact while the user types.
- **Round-trip preservation**: `normalizeHexValues` runs once
  after `parseYamlSectionValues` to convert numeric hex-typed
  values back to their canonical `"0x..."` string form. The YAML
  serializer emits strings verbatim (no quoting needed for
  `0x76`), so an unrelated edit on the same section now preserves
  the hex address instead of silently flipping it to decimal.

### Intentional canonicalisation behaviour

A YAML that already has the address written in **decimal** form
— e.g. `address: 119` instead of the more common `address: 0x77`
— gets canonicalised to hex on the next save of that section.
Both forms are valid ESPHome input and round-trip to the same
integer; the form normalises on load (so the user sees `0x77`
in the field), and the serializer emits the canonical hex on the
way back out.

This is a one-way upgrade — the integer value is identical, only
the on-disk notation changes — and it's the natural fallout of
the round-trip preservation logic above. Calling it out so a
reviewer doesn't think the YAML diff is accidental.

### Refactoring while we're in here

Extracted `renderFieldShell` to
`config-entry-renderers-shared.ts` — every primitive renderer
was repeating the `<div class="field"> + label + input + error`
envelope verbatim. `renderNumberField` and `renderHexIntField`
both consume the shared shell now; the long tail
(`renderStringField`, the password / suggestion-select
branches) can migrate as separate cleanup.

21 unit tests cover the helpers
(`test/util/hex-int.test.ts`) — hex/decimal parsing, edge cases
(`0x` with no digits, trailing junk, bare hex letters,
whitespace), the format filter (NaN/fractional/negative reject to
`""`), and the normaliser's identity-preservation shortcut for
non-hex sections.

**Related issue or feature (if applicable):**

- fixes esphome/device-builder#410

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue) — `bugfix`
- [ ] New feature (non-breaking change which adds functionality) — `new-feature`
- [ ] Enhancement to an existing feature — `enhancement`
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — `breaking-change`
- [ ] Refactor (no behaviour change) — `refactor`
- [ ] Documentation only — `docs`
- [ ] Maintenance / chore — `maintenance`
- [ ] CI / workflow change — `ci`
- [ ] Dependencies bump — `dependencies`

## Checklist

- [x] The code change is tested and works locally.
- [x] `npm run lint` passes.
- [x] `npm run test` passes.
- [x] Tests have been added to verify that the new code works (where applicable).

### CODE REVIEW

File: src/util/hex-int.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    if (entry.display_format !== "hex") continue;
+    const v = values[entry.key];
+    if (typeof v === "number") {
+      needsCopy = true;
+      break;
+    }
+  }
+  if (!needsCopy) return values;
+  const out: Record<string, unknown> = { ...values };
+  for (const entry of entries) {
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/hex-int.ts]: Real bug, fixed in ca0ebe5. The spread copy was promoting the null-prototype map back to a regular `Object`-prototype object, re-opening the prototype-pollution surface `parseYamlSectionValues` deliberately defends against. Switched to `Object.assign(Object.create(getPrototypeOf(values)), values)` so a null prototype survives the copy. Two new tests pin both the null-prototype and `Object`-prototype paths.

File: src/components/device/config-entry-renderers.ts [hunk 2]
```diff
@@ -57,31 +59,129 @@ export function renderNumberField(
...
+          ctx.emitChange(path, raw);
+          return;
+        }
+        // Emit the canonical ``"0x..."`` string so the YAML
+        // serializer writes hex (``address: 0x76``) instead of
+        // decimal — matches what gets normalised at parse time
+        // for fields the user didn't touch. ESPHome's
+        // ``cv.hex_int`` accepts the string form.
+        ctx.emitChange(path, formatHexInt(parsed));
+      }}
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/components/device/config-entry-renderers.ts]: Right — `-1` would round-trip as `emitChange("")`, same as the cleared-field branch, dropping the value from the YAML. Fixed in ca0ebe5: now if either `parseHexInt` returns `null` OR the parsed value's `formatHexInt` returns `""` (negative / NaN / fractional), we emit the raw string so the inline validator flags it instead of the form silently swallowing the input.

File: src/util/hex-int.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+ */
+
+/**
+ * Parse user-typed input into an integer.
+ *
+ * Accepts:
+ *  - hex with `0x` / `0X` prefix (`"0x76"`, `"0X1A"`);
+ *  - bare hex when the leading character disambiguates it as such
+ *    (intentionally NOT supported — `"76"` is decimal, see below);
+ *  - decimal (`"118"`);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/hex-int.ts]: Reworded in ca0ebe5 — the "Accepts" list now just covers the two real cases (`0x`-prefixed hex and decimal) and the "intentionally rejected" reasoning lives in its own paragraph below, with the bare-hex-letter case (`abc` / `ff`) called out explicitly so a reader doesn't wonder whether unprefixed hex letters get a different treatment from unprefixed digits.

File: src/util/hex-int.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+ *   "118")` → number. Both hex and decimal-looking input are
+ *   accepted; the user can type whichever is most natural for the
+ *   value at hand. Empty string → `null` (so optional entries get
+ *   stripped from the payload by the form's coerce pass).
+ *
+ * Round-trip preservation: when the YAML had `address: 0x76`, the
+ * form shows `0x76` and the user can save without forcing the
+ * file to flip to `address: 118`. The serializer in
+ * `yaml-section-values.ts` formats hex-typed values as `0x..`
+ * literals on write so the on-disk shape stays readable.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/hex-int.ts]: Fixed in ca0ebe5. The module header now spells out the actual mechanism: this module pre-formats hex-typed values into `"0x.."` strings (via `normalizeHexValues` at parse time and the renderer's emit-as-string at edit time); the YAML serializer is schema-agnostic, but it passes string scalars through verbatim, so a value already shaped as `"0x76"` lands on disk as `address: 0x76` without the serializer needing to know about the hex hint.

File: src/util/hex-int.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+ * edited. ESPHome's ``cv.hex_int`` validator accepts either form
+ * (``0x76`` unquoted, ``"0x76"`` quoted, or ``118`` decimal), so
+ * the on-disk shape stays valid either way.
+ *
+ * Top-level only — nested hex fields would need a recursive walk
+ * over ``entry.config_entries`` (i2c addresses are flat children
+ * of their component, no nesting today; revisit if a future hex
+ * field lands inside a NESTED group).
+ *
+ * Returns a fresh object; the input is never mutated.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/hex-int.ts]: Fixed in ca0ebe5 — the early-return identity-preservation is deliberate (cheap shortcut for non-hex sections, and callers may rely on the identity equality), so I updated the doc to describe both branches: "Returns the input object identity-equal when no rewrites are needed... otherwise returns a fresh object preserving the input's prototype."

File: src/util/hex-int.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+ * dropped" (intending `0x76` = 118). YAML and ESPHome both treat
+ * unprefixed input as decimal, so we follow that: typing `76`
+ * saves as `address: 76` (decimal), typing `0x76` saves as
+ * `address: 0x76`. Bare hex letters (`"abc"`, `"ff"`) hit
+ * neither regex and return `null`.
+ */
+export function parseHexInt(raw: string): number | null {
+  const trimmed = raw.trim();
+  if (trimmed === "") return null;
+  // Reject internal whitespace / sign / exponents. ``Number.parseInt``
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/hex-int.ts]: Right — the decimal regex `/^-?\d+$/` does allow a leading `-`, the comment was wrong. Fixed in ff8c0ad: reworded to enumerate what's actually rejected (internal whitespace, `+` sign, exponents, fractional, trailing characters, unprefixed hex letters).

File: src/util/hex-int.ts [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+/**
+ * Format an arbitrary form value as a hex literal for display.
+ *
+ * Returns `"0x" + lowercase-hex` matching ESPHome's own
+ * `cv.hex_int` formatter (`f"0x{value:02X}"` style — caps in
+ * Python, but the dashboard standardised on lowercase to match
+ * the `0x76` form Home Assistant docs and most i2c datasheets
+ * use).
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/hex-int.ts]: Right — the implementation uses `value.toString(16)` (minimum-width, no padding) so `0` → `0x0`, `0x76` → `0x76`. The Python format reference was misleading. Fixed in ff8c0ad: reworded to describe the actual no-padding behaviour and added the rationale (forcing a fixed width would mismatch the larger hex types — `hex_uint16_t`, `hex_uint32_t` — sharing the formatter).

File: src/util/hex-int.ts [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+ * `config-entry-renderers.ts`'s number-field branch. The native
+ * `<input type="number">` doesn't accept `0x...` literals, so the
+ * hex-display branch routes through these helpers instead.
+ *
+ * Two-way conversion shape:
+ *
+ * - **YAML → form display**: any integer (`118`, `0x76`, `"0x76"`)
+ *   resolves to a single number. We render it as
+ *   `formatHexInt(value)` → `"0x76"` (lowercase canonical form,
+ *   matching `repr()` and `cv.hex_int`'s own output formatter).
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/hex-int.ts]: Right — fixed in 530e684. The module header now describes the actual behaviour (lowercase, no padding, `value.toString(16)`-style) and explicitly notes the divergence from `cv.hex_int`'s uppercase output, with the rationale (HA docs and i2c datasheets use lowercase, so user-facing display follows that convention).

File: src/util/hex-int.ts [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+  const trimmed = raw.trim();
+  if (trimmed === "") return null;
+  // Strict regex gate. ``Number.parseInt`` happily eats trailing
+  // junk (``parseInt("0x76xyz") === 118``), which would silently
+  // swallow user typos. The two regexes accept only:
+  //   - ``0x`` / ``0X`` prefix followed by one or more hex digits;
+  //   - an optional leading ``-`` and one or more decimal digits.
+  // Everything else — internal whitespace, ``+`` sign, exponents
+  // (``1e3``), fractional (``3.14``), trailing characters,
+  // unprefixed hex letters — falls through to ``return null``.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/hex-int.ts:69]: The current comment already enumerates `` `+` sign `` explicitly (with the accept-list above noting "an optional leading `` `-` `` and one or more decimal digits"), so the rejection list isn't actually contradictory — Copilot misread. Leaving the comment as-is; happy to tweak if you want it more explicit.

File: src/util/hex-int.ts [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+ *
+ * Prototype preservation matters because ``parseYamlSectionValues``
+ * deliberately produces a null-prototype map to defend against
+ * user-keyed YAML containing ``__proto__`` / ``constructor`` /
+ * ``prototype`` (which would otherwise mutate the inherited
+ * prototype chain via ordinary property assignment). A naive
+ * ``{ ...values }`` spread re-opens that attack surface by
+ * promoting the result back to a regular ``Object``-prototype
+ * object; we copy into ``Object.create(null)`` instead so a
+ * crafted YAML can't escalate to prototype pollution.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/hex-int.ts]: Right, the doc was stale. Fixed in 530e684: now describes the actual cloning strategy (`Object.create(getPrototypeOf(values), getOwnPropertyDescriptors(values))`) which preserves whatever prototype the input had — null for the parser's defensive map, `Object.prototype` for plain test fixtures — and uses `defineProperty` semantics so a hostile `__proto__` source key can't trigger prototype mutation on the target. Kept the prototype-pollution rationale.

File: src/util/config-entry-defaults.ts [hunk 11]
```diff
@@ -30,6 +30,7 @@ export function makeConfigEntry(
     options: null,
     allow_custom_value: false,
     range: null,
+    display_format: null,
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/config-entry-defaults.ts:36]: Fixed in 530e684 — but instead of bumping the count to 29 (which would drift again the next time we add a field), removed the hardcoded number entirely and added a note that the count itself is the drift problem this helper is trying to solve. The type definition is the source of truth.
